### PR TITLE
Page for department

### PR DESCRIPTION
**Pull request by YuetingLiao**
**Issue**
Hello Everyone,
This pull request update features on pages.
User will be redirected to a subject page Instead of using accordion for departments in Course Explorer.
All courses for a subject will be displayed into two columns.
Also, a new page is created for course detail.

*Formal Issue:*
This Pull Request still works on issue #191/ #192

**Database Changes/Migrations**
N/A

**Test Modifications**
N/A

**Test Procedure**

Show procedure to test functionality with a clear procedure

*Test New Subject Page:*
> 1. Load Spring 2020 CSV
> 2. Navigates to the course explore page by click on the blue tag "explore"
> 3. Click "CSCI" subject under the Science department
> 4. The new subject page should be displayed on http://localhost:8080/CSCI. All courses are listed "evenly" by two columns

*Test New Course Page:*
> 1. Access CSCI subject page on http://localhost:8080/CSCI.
> 2. Click "Computer Science 1 CSCI-1100" button
> 3. Click "CSCI" subject under the Science department
> 4. The new course page should be displayed on http://localhost:8080/CSCI/CSCI-1100. Detail course information should be displayed.

**Photos**
Show before and after, capture screenshot/gif of finished feature/bug
4d63927f8af5b44567410b9f67d8599

0897532793b78fae8b08d34cece6a3e

**Additional Info**
This Pull Request also includes the contribution of Ike Lyons, we are using the same branch.

### CODE REVIEW

**Review by 123joshuawu — CHANGES_REQUESTED** [review]
Nice! Good stuff here, left some comments

**Comment by 123joshuawu** [comment]
So which issue is this PR for? The title seems to indicate #191 but looks like #192 is also in here

File: src/web/src/pages/ManageAccounts.vue [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<template>
```
Comment (123joshuawu) [inline src/web/src/pages/ManageAccounts.vue:1]: This should be separate PR

File: src/web/src/pages/Admin.vue [hunk 2]
```diff
@@ -32,10 +32,14 @@
       <SetDefault />
```
Comment (123joshuawu) [inline src/web/src/pages/Admin.vue:32]: Separate PR

File: src/api/controller/user.py [hunk 3]
```diff
@@ -132,9 +132,15 @@ def add_user(form):
     major = form['major']
     degree = form['degree']
 
+    if name.strip() == "":
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/api/controller/user.py:135]: Got it

File: src/web/src/components/DepartmentList.vue [hunk 4]
```diff
@@ -22,7 +33,7 @@
                   <li
                     class="course-listing"
                     v-for="course in deptClassDict[major]"
-                    :key="course.name + course.date_start + course.date_end"
+                    :key="course.name + course.date_start + course.date_start"
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/components/DepartmentList.vue:36]: I'm changing it back. This is a mistake I made when I was testing my code, thank you for pointing it out.

File: src/web/src/components/Header.vue [hunk 5]
```diff
@@ -5,14 +5,10 @@
       <div class="semester">{{ selectedSemester }}</div>
       <b-navbar-nav>
         <b-nav-item>
-          <router-link :to="{ name: 'CourseScheduler' }">
-            Schedule
-          </router-link>
+          <router-link :to="{ name: 'CourseScheduler' }">Schedule</router-link>
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/components/Header.vue:8]: I believe it's prettier formatter thing ...

File: src/web/src/pages/CourseExplorer.vue [hunk 6]
```diff
@@ -148,7 +175,12 @@ export default {
     },
   },
   computed: {
-    //Used to define the placement of each department into the two main columns
+    /**
+     *coursesChunked
+     * Used to define the placement of each department into the two main columns
+     * Data structure -> An Array contains 6 list, each list contains one string
+     * @returns [{String}]
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/pages/CourseExplorer.vue:182]: Thank you a lot, I was about to ask that.

File: src/web/src/pages/CourseExplorer.vue [hunk 7]
```diff
@@ -125,6 +145,13 @@ export default {
     };
   },
   async created() {
+    if (this.selectedSemester === "") {
+      const querySemester = this.$route.query.semester;
+      this.selectedSemester =
+        querySemester && querySemester != "null"
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/pages/CourseExplorer.vue:151]: > So which issue is this PR for? The title seems to indicate #191 but looks like #192 is also in here

Yes, Ike and I are working on the same branch for #191 and #192 as we discussed before with Jacob. We thought this would make both of us easier.

**Comment by YuetingLiao** [comment]
> So which issue is this PR for? The title seems to indicate #191 but looks like #192 is also in here

**Comment by IkeLyons** [comment]
@123joshuawu,  @jshom wanted us to leave a pull request open for the two course explorer issues, #191 and #192 as their scope is similar. Were going to leave this open as we continue to commit so that people can see what its looking like and give feedback.

**Comment by 123joshuawu** [comment]
Gotcha, I'll ignore the CoursePage then

**Review by 123joshuawu — CHANGES_REQUESTED** [review]
Some fixes for `SubjectExplorer.vue`

File: src/web/src/pages/SubjectExplorer.vue [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+  /**
+   * created function calls automatically once the page is access/ object is created
+   * Loop through all courses in data base
+   * Only store the courses within the same subject/major into an array
+   * subjectCourseArr is an array of course objects
+   */
+  async created() {
+    getCourses(this.selectedSemester).then((courses) => {
+      for (const c of courses) {
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/pages/SubjectExplorer.vue:144]: Ty, `filter` is used instead of `for `

File: src/web/src/pages/SubjectExplorer.vue [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        if (c.department === this.subject) {
+          this.subjectCourseArr.push(c);
+        }
+      }
+      this.leftColumnCourseNum = Math.ceil(this.subjectCourseArr.length / 2);
```
[the remark this answers is not in the document]
Comment (YuetingLiao) [reply src/web/src/pages/SubjectExplorer.vue:149]: Great, I already refactored this part :)

File: src/web/src/pages/CourseExplorer.vue [hunk 10]
```diff
@@ -148,14 +161,35 @@ export default {
...
-        chunkedArr.push(this.schoolsMajorDict[this.schoolOrder[i]]);
+      let majorsArr = Object.values(this.schoolsMajorDict);
+      majorsArr.sort((a, b) => b.size - a.size);
+      for (let i = 0; i < 6; i++) {
+        chunkedArr.push(majorsArr[i]);
       }
       return chunkedArr;
     },
+
+    schoolOrder() {
```
Comment (123joshuawu) [inline src/web/src/pages/CourseExplorer.vue:180]: An alternative approach to having two computed methods `coursesChunked` and `schoolOrder` is to have a single computed method that **returns an array of objects**, each object containing a school and its majors. 

So something like this:
```javascript
computed: {
  schoolDepartmentObjects () {
    // Object.entries returns a list of key-value pairs
    return Object.entries(this.schoolsMajorDict)
                   // convert the lists into an object so they are easier to use
                   // [school, majors] -> {school, majors}
                   .map((schoolDepartmentMapping) => ({
                      school: schoolDepartmentMapping[0],
                      departments: schoolDepartmentMapping[1],
                   })
                   // sort objects by size of departments
                   .sort((left, right) => right.departments.size - left.departments.size); 
  }
}
```

This way, in the html, instead of doing a`v-for` on a number range and then calling `{{ schoolOrders[n-1] }}`, you can do something like

```vue
<div v-for="schoolDepartmentObj in schoolDepartmentObjects" :key="schoolDepartmentObj.school">
  ...
  {{ schoolDepartmentObj.school }}
  ...
  <div v-for="department in schoolDepartmentObj.departments">
    {{ department }}
  </div>
  ...
</div>
```

Generally, try to use arrays of objects and iterate over objects instead of using the indexes

File: src/web/src/pages/CoursePage.vue [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      let treeData = {};
+      treeData.name = this.courseName;
+      return treeData;
+    },
+  },
+};
+</script>
+
+<style>
+.container {
```
Comment (123joshuawu) [inline src/web/src/pages/CoursePage.vue:84]: if this is the only style for this, would prefer adding the class `text-center` instead

File: src/web/src/pages/CoursePage.vue [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  methods: {
+    generateRequirementsText,
+  },
+  async created() {
+    const querySemester = this.$route.query.semester;
+    this.selectedSemester =
+      querySemester && querySemester != "null"
+        ? querySemester
+        : await getDefaultSemester();
+    getCourses(this.selectedSemester).then((courses) => {
```
Comment (123joshuawu) [inline src/web/src/pages/CoursePage.vue:61]: `created` is async so you can use `await` syntax here

`const courses = await getCourses(this.selectedSemester)`

File: src/web/src/pages/CoursePage.vue [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+      for (const c of courses) {
+        this.allCoursePreReqs[c.name] = c.prerequisites;
```
Comment (123joshuawu) [inline src/web/src/pages/CoursePage.vue:63]: This should be relegated to backend, probably  a `/course/:id` route, but this is fine for now

**Review by 123joshuawu — CHANGES_REQUESTED** [review]
Sweet, some code changes and a design suggestion, but otherwise good.

Am looking forward to these

File: src/web/src/pages/CoursePage.vue (hunk 11 above)
Comment (IkeLyons) [reply src/web/src/pages/CoursePage.vue:84]: That class doesn't seem to work for some reason? Now that I'm looking at it left justified again I think I prefer it anyway. I'll delete the style in favor of having it left justified instead of centered.

File: src/web/src/pages/CourseExplorer.vue [hunk 14]
```diff
@@ -1,12 +1,29 @@
...
-  <div v-if="ready" class="gridContainer w-100 mb-4">
+  <div v-if="ready" class="gridContainer w-100">
     <b-row>
+      <!--
+        - Left side of the column
+        - Displays the first 2 departments in the schoolOrder (Science" and "Engineering")
+        - n starts with 1
+      -->
       <b-col>
         <b-row v-for="n in 2" :key="n" class="departmentBox border m-2 mb-4">
```
[the remark this answers is not in the document]
Comment (IkeLyons) [reply src/web/src/pages/CourseExplorer.vue:8]: That's what I though as well but without the key bound you get an error "Custom elements in iteration require 'v-bind:key' directives". Not sure why this is as the `n` in `n in 2` isn't a custom element but adding the key silences the error.

[the remark this answers is not in the document]
Comment (123joshuawu) [reply src/web/src/pages/CourseExplorer.vue:8]: Oof okay, that's algright

File: src/web/src/routes.js [hunk 15]
```diff
@@ -48,6 +62,11 @@ var router = new VueRouter({
       component: EditSemestersPage,
       name: "EditSemesters",
     },
+    {
+      path: "/admin/manageaccounts",
+      component: ManageUserAccountPage,
+      name: "ManageAccounts",
+    },
```
Comment (jshom) [inline src/web/src/routes.js:69]: Please remove this Manage User Account section for now, you can stash the changes or create another branch with the code.

Besides that should be good for merge

**Review by jshom — CHANGES_REQUESTED** [review]

File: src/web/src/routes.js (hunk 15 above)
Comment (YuetingLiao) [reply src/web/src/routes.js]: Thank you, I already removed all sections associated with Manage Account

**Comment by venkatrsrinivas** [comment]
Hello. 
This all looks really great to me!
I just had a Few Minor Comments. 

First and foremost, I think that the Title of the Department, like CSCI could potentially be moved down slightly.
Examining the Department Page on Dark Mode, it looks a little hard to make it out. 
I've attached a few Screenshots below. 

Screen Shot 2020-10-27 at 17 18 38

Aside from this, I noticed that when YACS was NOT on Dark Mode and you hover over the Courses, it looks like the Course is highlighted grey, which is great, but this similar thing is NOT present in the Dark Mode and makes it slightly hard to see as well. 

Screen Shot 2020-10-27 at 17 18 30

Once again, I'm not sure if this could be moved to another Issues, but figured it would not hurt to bring it up as they are just Minor, Minor Things.

Overall, I could not find anything wrong about the Content behind Department Page itself. 
Everything looks great!
Looking forward to seeing this in the Master Branch.

**Review by venkatrsrinivas — COMMENTED** [review]
Hello. 
This all looks really great to me!
I just had a Few Minor Comments. 

First and foremost, I think that the Title of the Department, like CSCI could potentially be moved down slightly.
Examining the Department Page on Dark Mode, it looks a little hard to make it out. 
I've attached a few Screenshots below. 

Screen Shot 2020-10-27 at 17 18 38

Aside from this, I noticed that when YACS was NOT on Dark Mode and you hover over the Courses, it looks like the Course is highlighted grey, which is great, but this similar thing is NOT present in the Dark Mode and makes it slightly hard to see as well. 

Screen Shot 2020-10-27 at 17 18 30

Once again, I'm not sure if this could be moved to other Issues, but figured it would not hurt to bring it up as they are just Minor, Minor Things.

Overall, I could not find anything wrong about the Content behind Department Page itself. 
Everything looks great!
Looking forward to seeing this in the Master Branch.

File: src/web/src/pages/Admin.vue [hunk 16]
```diff
@@ -32,10 +32,14 @@
...
     </b-modal>
 
+    <router-link :to="{ name: 'ManageAccounts' }">
+      Manage Accounts
+    </router-link>
```
Comment (jshom) [inline src/web/src/pages/Admin.vue:37]: Remove for now

**Comment by jshom** [comment]
> Hello.
> This all looks really great to me!
> I just had a Few Minor Comments.
> 
> First and foremost, I think that the Title of the Department, like CSCI could potentially be moved down slightly.
> Examining the Department Page on Dark Mode, it looks a little hard to make it out.
> I've attached a few Screenshots below.
> 
> Screen Shot 2020-10-27 at 17 18 38
> 
> Aside from this, I noticed that when YACS was NOT on Dark Mode and you hover over the Courses, it looks like the Course is highlighted grey, which is great, but this similar thing is NOT present in the Dark Mode and makes it slightly hard to see as well.
> 
> Screen Shot 2020-10-27 at 17 18 30
> 
> Once again, I'm not sure if this could be moved to other Issues, but figured it would not hurt to bring it up as they are just Minor, Minor Things.
> 
> Overall, I could not find anything wrong about the Content behind Department Page itself.
> Everything looks great!
> Looking forward to seeing this in the Master Branch.

Yes, we'll be adding the visual changes in to a new issue as this PR is getting a bit too big. Will add this comment to the new issue once created.

Thanks for the screenshots and requests!

**Review by 123joshuawu — APPROVED** [review]

**Review by venkatrsrinivas — APPROVED** [review]

**Review by jshom — APPROVED** [review]
